### ISSUE TITLE
Weird code for _mm_cvtepi32_ps on haswell/skylake

### ISSUE THREAD

**Issue by bartek-siudeja** [AuburnSounds/intel-intrinsics#86]
First, amazing library! Big thanks for doing this.

I had to save some partial vectors after integer to float conversion, and I noticed ldc going a bit crazy. 

```
float4 _mm_cvtepi32_ps(int4 a)
{
    float4 res;
    static foreach(i; 0..4) // 0..3 gives the same code in save_f4
        res[i] = a[i];
    return res;
    // return __asm!float4("vcvtdq2ps $1, $0", "=x, x", a);
}

void save_f4(ref float[3] arr, int4 i4)
{
    float4 f4 = _mm_cvtepi32_ps(i4);
    f4 *= 1.23;
    arr[0..$] = f4.array[0..arr.length];
}
```
It seems ldc is trying to handle each of the three floats separately, but it is using `vcvtdq2ps` on each of them. From godbolt:
```
void example.save_f4(ref float[3], __vector(int[4])):
        vcvtdq2ps       xmm1, xmm0
        // unnecessary instructions
        vpermilps       xmm2, xmm0, 85
        vcvtdq2ps       xmm2, xmm2
        vunpcklps       xmm1, xmm1, xmm2
        vpermilps       xmm0, xmm0, 238
        vcvtdq2ps       xmm0, xmm0
        vinsertps       xmm0, xmm1, xmm0, 32
        // end unnecessary
        vbroadcastss    xmm1, dword ptr [rip + .LCPI1_0]
        vmulps  xmm0, xmm0, xmm1
        vextractps      dword ptr [rdi + 8], xmm0, 2
        vmovlps qword ptr [rdi], xmm0
        ret
```
Maybe this is not avoidable without IR or assembly.

**Comment by p0nce** [comment]
Hello, could you a bit clearer for what actions and steps you need from me? Is there a bug in `intel-intrinsics`?

**Comment by bartek-siudeja** [comment]
Sorry, it seems that `_mm_cvtepi32_ps` as defined in `intel-intrinsics` converts to `vcvtdq2ps` in normal cases. But compiler is trying too hard to optimize inside the D code if it discovers that not all entries of the vector are used by subsequent code. I am wondering if there is a way to prevent ldc from "optimizing", looking inside this intrinsic.

Eg. `__asm!float4("vcvtdq2ps $1, $0", "=x, x", a)` as implementation will stop ldc from inserting unnecessary vpermilps/vunpcklps/vinsertps instructions. Perhaps there is a way to improve this intrinsic.

**Comment by p0nce** [comment]
Fixed in v1.6.2
Could reproduce with or without AVX.
Please note that this is clearly a LLVM performance issue, so it would be extra nice if you submit the bug to LLVM! Probalby exposed in clang also.

About asm inline: it's almost always bad, since giving the exaxct instruction will force VEX or non-VEX hence harmful transitions. LLVM IR is fine for those cases.

**Comment by p0nce** [comment]
Checked that in clang and no it doesn't really expose the same bug.
